### PR TITLE
#1830 implement {DataFrame, Series}.empty

### PR DESCRIPTION
**Pull request by Fernadoo**
## What do these changes do?

I implemented .empty property for {DataFrame, Series}.

## Related issue number

Issue #1830 should be resolved

### CODE REVIEW

**Comment by hekaisheng** [comment]
Thanks for your contribution! Could you add test cases in `dataframe.base.tests` ?

**Comment by Fernadoo** [comment]
> Thanks for your contribution! Could you add test cases in `dataframe.base.tests` ?

Sure, I am gonna do it later today :D

**Comment by hekaisheng** [comment]
> > Thanks for your contribution! Could you add test cases in `dataframe.base.tests` ?
> 
> Sure, I am gonna do it later today :D

BTW, we merged a large pull request, please rebase it first and adapt to the new style of tests.

**Comment by Fernadoo** [comment]
> > > Thanks for your contribution! Could you add test cases in `dataframe.base.tests` ?
> > 
> > 
> > Sure, I am gonna do it later today :D
> 
> BTW, we merged a large pull request, please rebase it first and adapt to the new style of tests.

Excuse me, I got a bit confused. The commit history shows that you have merged the latest PR into v0.6. Nevertheless, according to the contributing guideline, I was developing upon master and submitted my PR against master as well. So do you mean that I am indeed supposed to rebase v0.6 to my private branch and also develop upon v0.6?

**Comment by qinxuye** [comment]
> > > > Thanks for your contribution! Could you add test cases in `dataframe.base.tests` ?
> > > 
> > > 
> > > Sure, I am gonna do it later today :D
> > 
> > 
> > BTW, we merged a large pull request, please rebase it first and adapt to the new style of tests.
> 
> Excuse me, I got a bit confused. The commit history shows that you have merged the latest PR into v0.6. Nevertheless, according to the contributing guideline, I was developing upon master and submitted my PR against master as well. So do you mean that I am indeed supposed to rebase v0.6 to my private branch and also develop upon v0.6?

@hekaisheng means commit https://github.com/mars-project/mars/commit/37455c1007ca1d255de9ca789de5da70cbf1864e, just rebase the master branch please.

**Comment by Fernadoo** [comment]
> > > > > Thanks for your contribution! Could you add test cases in `dataframe.base.tests` ?
> > > > 
> > > > 
> > > > Sure, I am gonna do it later today :D
> > > 
> > > 
> > > BTW, we merged a large pull request, please rebase it first and adapt to the new style of tests.
> > 
> > 
> > Excuse me, I got a bit confused. The commit history shows that you have merged the latest PR into v0.6. Nevertheless, according to the contributing guideline, I was developing upon master and submitted my PR against master as well. So do you mean that I am indeed supposed to rebase v0.6 to my private branch and also develop upon v0.6?
> 
> @hekaisheng means commit [37455c1](https://github.com/mars-project/mars/commit/37455c1007ca1d255de9ca789de5da70cbf1864e), just rebase the master branch please.

Oh, I see! Sure then.

**Comment by Fernadoo** [comment]
> > > > > > Thanks for your contribution! Could you add test cases in `dataframe.base.tests` ?
> > > > > 
> > > > > 
> > > > > Sure, I am gonna do it later today :D
> > > > 
> > > > 
> > > > BTW, we merged a large pull request, please rebase it first and adapt to the new style of tests.
> > > 
> > > 
> > > Excuse me, I got a bit confused. The commit history shows that you have merged the latest PR into v0.6. Nevertheless, according to the contributing guideline, I was developing upon master and submitted my PR against master as well. So do you mean that I am indeed supposed to rebase v0.6 to my private branch and also develop upon v0.6?
> > 
> > 
> > @hekaisheng means commit [37455c1](https://github.com/mars-project/mars/commit/37455c1007ca1d255de9ca789de5da70cbf1864e), just rebase the master branch please.
> 
> Oh, I see! Sure then.

Since `pytest mars` did not work well under my forked repo, so I just downloaded [`mars-project:master`](https://github.com/mars-project/mars) without any modification to evaluate all test cases. It is supposed to pass all cases. However, it turns out that the command `pytest mars` will always be stuck in `mars/dataframe/arithmetic/tests/test_arithmetic_execution.py`, attached as the figure below (sometimes even outputs an 'F' and stops moving on):
pytest_err
I am wondering if this situation has happened before?

**Comment by hekaisheng** [comment]
Could you try to run `python setup.py build_ext -i` and run tests again?  It will takes a long time to check all cases, just make sure the cases you added are pass.

**Comment by qinxuye** [comment]
@hekaisheng is right, you just need to pay attention to the test you added, the entire unittests are too huge to cover.

**Comment by Fernadoo** [comment]
Hi @hekaisheng and @qinxuye, I have added some test cases in `mars/dataframe/base/tests/test_base.py`. The pytest results show that those newly added cases are passed. Is there anything left that I can help with? 
(BTW, I have already rebased upon your latest commit [`2805241`](https://github.com/mars-project/mars/commit/2805241ac55b50c4f6319baa41113fbf8c723832), no worries)

**Comment by hekaisheng** [comment]
Could you test with this case:
``` Python
a = md.DataFrame(np.random.rand(10, 2))
assert a[a > 0].empty
```

In some cases, Mars tensor or DataFrame has unknown shape as Mars is a lazy evaluation framework, its shape can't be decided before execution.

File: mars/dataframe/core.py [hunk 1]
```diff
@@ -1021,6 +1021,18 @@ def index(self):
     def axes(self):
         return [self.index]
 
+    """
+    Property .empty STARTS
```
Comment (hekaisheng) [inline mars/dataframe/core.py:1025]: Usually, we use git blame to show the the revision and author of code, comments are not necessary.

**Comment by Fernadoo** [comment]
> Could you test with this case:
> 
> ```python
> a = md.DataFrame(np.random.rand(10, 2))
> assert a[a > 0].empty
> ```
> 
> In some cases, Mars tensor or DataFrame has unknown shape as Mars is a lazy evaluation framework, its shape can't be decided before execution.

Your conjecture is right, `a[a > 0].empty` does not return `True`. Let me fix this.

File: mars/dataframe/core.py (hunk 1 above)
Comment (Fernadoo) [reply mars/dataframe/core.py:1025]: Sure then :D

**Comment by hekaisheng** [comment]
We implement `mt.shape` (mars/tensor.shape) to get shape of those tensors has unknown shape, if mars.dataframe.shape is implemented, `df.empty` can return a Mars tensor and get value after triggering execution.

**Comment by hekaisheng** [comment]
For this PR, you can just raise an error if DataFrame's shape has `np.nan`, and tell users to trigger execution first, its shape will be updated after call `df.execute()`.

**Comment by Fernadoo** [comment]
Hi @hekaisheng, I have added `raise` module and deleted unnecessary comments. Plz have a look:D

File: mars/dataframe/base/tests/test_base.py [hunk 2]
```diff
@@ -888,3 +888,21 @@ def a():
...
+    assert from_pandas_df(pd.DataFrame({'a':[1], 'b':[2]})).empty == pd.DataFrame({'a':[1], 'b':[2]}).empty
+
+    # for Series
+    assert from_pandas_series(pd.Series()).empty == pd.Series().empty
+    assert from_pandas_series(pd.Series({})).empty == pd.Series({}).empty
+    assert from_pandas_series(pd.Series({'a':[]})).empty == pd.Series({'a':[]}).empty
+    assert from_pandas_series(pd.Series({'a':[1]})).empty == pd.Series({'a':[1]}).empty
+
+    # Maybe fail due to lazy evaluation
+    # assert from_pandas_df(pd.DataFrame(np.random.rand(10, 2))).empty == True
```
Comment (hekaisheng) [inline mars/dataframe/base/tests/test_base.py:908]: Use `pytest.raises` to test unknown shape.

File: mars/dataframe/core.py [hunk 3]
```diff
@@ -1499,6 +1506,13 @@ def columns(self):
     def axes(self):
         return [self.index, self.columns]
 
+    @property
+    def empty(self):
+        shape = getattr(self, '_shape')
+        if np.nan in shape:
+            raise RuntimeError("The shape is unknown until you trigger execution")
+        return shape[0] == 0
```
Comment (hekaisheng) [inline mars/dataframe/core.py:1514]: if length of columns is 0, `df.empty` also returns True

**Review by hekaisheng — CHANGES_REQUESTED** [review]

File: mars/dataframe/base/tests/test_base.py (hunk 2 above)
Comment (Fernadoo) [reply mars/dataframe/base/tests/test_base.py]: Resolved!

File: mars/dataframe/core.py (hunk 3 above)
Comment (Fernadoo) [reply mars/dataframe/core.py]: Resolved

**Comment by Fernadoo** [comment]
Please have a look if it's ok now :D, thanks for your patience!

File: mars/dataframe/base/tests/test_base.py [hunk 4]
```diff
@@ -888,3 +888,24 @@ def a():
...
+    # for Series
+    assert from_pandas_series(pd.Series()).empty == pd.Series().empty
+    assert from_pandas_series(pd.Series({})).empty == pd.Series({}).empty
+    assert from_pandas_series(pd.Series({'a':[]})).empty == pd.Series({'a':[]}).empty
+    assert from_pandas_series(pd.Series({'a':[1]})).empty == pd.Series({'a':[1]}).empty
+
+    # Maybe fail due to lazy evaluation
+    with pytest.raises(RuntimeError):
+        a = from_pandas_df(pd.DataFrame(np.random.rand(10, 2)))
+        assert a[a>0].empty == True
```
Comment (hekaisheng) [inline mars/dataframe/base/tests/test_base.py:911]: Add a new line at end of file

Comment (Fernadoo) [reply mars/dataframe/base/tests/test_base.py]: Have done it but don't know why we need it

Comment (Fernadoo) [reply mars/dataframe/base/tests/test_base.py]: It seems that I also need to modify a[a>0] to a[a > 0], because of [ubuntu-latest, 3.9-cython] tests?

Comment (hekaisheng) [reply mars/dataframe/base/tests/test_base.py]: We use Flake8 for code linting, it breaks the rule `No newline at end of file (W292)`.

Comment (Fernadoo) [reply mars/dataframe/base/tests/test_base.py]: Then let me fix this and `reset` to make the log a bit cleaner

File: mars/dataframe/core.py [hunk 5]
```diff
@@ -1021,6 +1021,13 @@ def index(self):
     def axes(self):
         return [self.index]
 
+    @property
+    def empty(self):
+        shape = getattr(self, '_shape')
+        if np.nan in shape:
+            raise RuntimeError("The shape is unknown until you trigger execution")
```
Comment (qinxuye) [inline mars/dataframe/core.py:1028]: `ValueError` can be raised which can refer to https://github.com/mars-project/mars/blob/6535fe1b198e2ef7b209058d442a3b3f375cfba5/mars/core/entity/executable.py#L90-L91 .

File: mars/dataframe/core.py [hunk 6]
```diff
@@ -1499,6 +1506,13 @@ def columns(self):
     def axes(self):
         return [self.index, self.columns]
 
+    @property
+    def empty(self):
+        shape = getattr(self, '_shape')
+        if np.nan in shape:
```
Comment (hekaisheng) [inline mars/dataframe/core.py:1512]: It's better to use `np.isnan` to check.

**Review by qinxuye — COMMENTED** [review]
Code in `actors` directory need to be removed.

**Review by hekaisheng — CHANGES_REQUESTED** [review]
Could you check your changes to make sure other files are not included?

**Comment by Fernadoo** [comment]
@hekaisheng @qinxuye, sure, also notice that, don't know why but I shall fix these

**Comment by Fernadoo** [comment]
New implementation redirected to [#2155](https://github.com/mars-project/mars/pull/2155)

**Pull request closed without merging**